staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go

```diff
@@ -161,8 +166,42 @@ func prepareEnvSet(baseEnvSet *environment.EnvSet, declType *apiservercel.DeclTy
 			EnvOptions: []cel.EnvOption{
 				cel.Variable(OldScopedVarName, scopedType.CelType()),
 			},
+			DeclTypes: []*apiservercel.DeclType{
+				scopedType,
+			},
 		},
 	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	optionalOldSelfEnvSet, err := baseEnvSet.Extend(
+		environment.VersionedOptions{
+			// Feature epoch was actually 1.23, but we artificially set it to 1.0 because these
+			// options should always be present.
+			IntroducedVersion: version.MajorMinor(1, 0),
+			EnvOptions: []cel.EnvOption{
+				cel.Variable(ScopedVarName, scopedType.CelType()),
+			},
+			DeclTypes: []*apiservercel.DeclType{
+				scopedType,
+			},
+		},
+		environment.VersionedOptions{
+			IntroducedVersion: version.MajorMinor(1, 24),
+			EnvOptions: []cel.EnvOption{
+				cel.Variable(OldScopedVarName, types.NewOptionalType(scopedType.CelType())),
+			},
+			DeclTypes: []*apiservercel.DeclType{
+				scopedType,
+			},
```

Comment:
Curious if this was strictly needed since the decl is already added to the environment in the prior version options?  If we don't need to declare it again maybe omit it?  (if we do need to declare it again, I'll want to dig in to figure out why..)

Reply:
Was able to remove it. It needed to be added in an earlier implementation when I had called extend multiple times like the following pseudocode:

```go
selfEnvSet := baseEnv.Extend({
     // self definition
    DeclTypes: [scopedType]
})

oldSelfEnvSet := selfEnvSet.Extend({
  // oldself definition
})


optionalOldSelfEnvSet := selfEnvSet.Extend({
    // optionalOldSelf definition
})
```

This caused field resolution errors that weren't resolved until we either added the DeclType to each call to Extend, or made it the form it is currently.

@jpbetz is it expected that the definition wouldn't be available in successive calls to `Extend`?